src/main/java/org/elmlang/intellijplugin/features/intention/imports/ElmAddImportHelper.java

```diff
@@ -0,0 +1,199 @@
+package org.elmlang.intellijplugin.features.intention.imports;
+
+import com.intellij.lang.ASTNode;
+import com.intellij.openapi.project.Project;
+import com.intellij.psi.PsiComment;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.util.PsiTreeUtil;
+import org.elmlang.intellijplugin.psi.*;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class ElmAddImportHelper {
+    private ElmAddImportHelper() {
+    }
+
+    public static void addOrUpdateImport(ElmFile sourceFile, String moduleName, @Nullable String nameToImport, boolean importAsQualified) {
+        Project project = sourceFile.getProject();
+
+        // Create a new, stand-alone import based solely on the quick-fix
+        ElmImportClause newImport = importAsQualified
+                ? ElmElementFactory.createImport(project, moduleName)
+                : ElmElementFactory.createImportExposing(project, moduleName, nameToImport);
+
+        // If there are any existing imports for the same module, merge with it
+        Optional<ElmImportClause> existingImportClause = sourceFile.getImportClauseByModuleName(moduleName);
+        if (existingImportClause.isPresent()) {
+            // merge with existing import
+            ElmImportClause oldImport = existingImportClause.get();
+            ElmImportClause mergedImport = mergeImports(sourceFile, oldImport, newImport);
+            oldImport.replace(mergedImport);
+        } else {
+            // insert a new import clause
+            ASTNode insertPosition = getInsertPosition(sourceFile, moduleName);
+            insertImportClause(newImport, insertPosition);
+        }
+    }
+
+    private static ElmImportClause mergeImports(ElmFile sourceFile, ElmImportClause import1, ElmImportClause import2) {
+        assert Objects.equals(import1.getModuleName().getText(),
+                              import2.getModuleName().getText());
+
+        Project project = sourceFile.getProject();
+
+        // If either one of the imports has an alias clause, extract it
+        // for use in the new, merged import.
+        Optional<ElmAsClause> as1 = Optional.ofNullable(import1.getAsClause());
+        Optional<ElmAsClause> as2 = Optional.ofNullable(import2.getAsClause());
+        String aliasText = Optional.ofNullable(as1.orElse(as2.orElse(null)))
+                .map(e -> " as " + e.getUpperCaseId().getName())
+                .orElse("");
+
+        // Merge the exposing clause(s)
+        Optional<ElmExposingClause> exposing1 = Optional.ofNullable(import1.getExposingClause());
+        Optional<ElmExposingClause> exposing2 = Optional.ofNullable(import2.getExposingClause());
+
+        List<String> exposedValues =
```

Comment:
You don't need the stream to be converted to a list at this point, do you? You can use `Stream.concat` to merge it with `exposedTypes`, then sort the result stream and then convert it to a list if needed.